### PR TITLE
Delete threat intel source config API

### PR DESCRIPTION
**Pull request by jowg-amazon**
### Description
Architecture for delete threat intel source config API
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigRequest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public String getId() {
+        return id;
+    }
+
+
+    @Override
+    public ActionRequestValidationException validate() {
+        return null;
```
Comment (AWSHurneyt) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigRequest.java:44]: Should there be a check here to assert that the `id` is not null or empty?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigManagementService.java [hunk 2]
```diff
@@ -158,7 +159,8 @@ public void deleteTIFSourceConfig(
                     SATIFSourceConfig SaTifSourceConfig = convertToSATIFConfig(SaTifSourceConfigDto);
                     SaTifSourceConfigService.deleteTIFSourceConfig(SaTifSourceConfig, ActionListener.wrap(
                             deleteResponse -> {
-                                log.debug("Successfully deleted threat intel source config");
+                                log.debug("Successfully deleted threat intel source config [{}]", SaTifSourceConfig.getId());
```
Comment (AWSHurneyt) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigManagementService.java:162]: Not blocking, but should we consider making this an info-level log?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java [hunk 3]
```diff
@@ -248,7 +246,7 @@ public void deleteTIFSourceConfig(
         client.delete(request, ActionListener.wrap(
                 deleteResponse -> {
                     if (deleteResponse.status().equals(RestStatus.OK)) {
-                        log.info("Deleted threat intel source config [{}] successfully", SaTifSourceConfig.getId());
+                        log.debug("Deleted threat intel source config [{}] successfully", SaTifSourceConfig.getId());
```
Comment (AWSHurneyt) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java:249]: Same question as above.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigRequest.java (hunk 1 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigRequest.java]: Added a check to ensure id exists

**Review by AWSHurneyt — APPROVED** [review]

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigResponse.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.opensearch.securityanalytics.threatIntel.model.SATIFSourceConfigDto;
+
+import java.io.IOException;
+
+import static org.opensearch.securityanalytics.util.RestHandlerUtils._ID;
+import static org.opensearch.securityanalytics.util.RestHandlerUtils._VERSION;
+
+public class SADeleteTIFSourceConfigResponse extends ActionResponse implements ToXContentObject {
+    private final String id;
+    private final RestStatus status;
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigResponse.java:22]: what status implies deleted?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/common/Constants.java [hunk 5]
```diff
@@ -10,4 +10,6 @@
 public class Constants {
     public static final String USER_AGENT_KEY = "User-Agent";
     public static final String USER_AGENT_VALUE = String.format(Locale.ROOT, "OpenSearch/%s vanilla", Version.CURRENT.toString());
+    public static final String SOURCE_CONFIG_ID = "source_config_id";
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/common/Constants.java:13]: MEGA-NIT: threat_intel_source_config_id?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(RestRequest.Method.DELETE, String.format(Locale.getDefault(), "%s/{%s}", SecurityAnalyticsPlugin.THREAT_INTEL_SOURCE_URI, SOURCE_CONFIG_ID)));
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        String SaTifSourceConfigId = request.param(SOURCE_CONFIG_ID, SATIFSourceConfigDto.NO_ID);
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java:36]: lower case letter to begin variable name

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        if (SaTifSourceConfigId == null || SaTifSourceConfigId.isEmpty()) {
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java:38]: nit: Stringutils.isBlank() instead

**Review by eirsep — CHANGES_REQUESTED** [review]
if there is only one config and it it's being deleted we need to ensure there are no jobs
Fail the delete request of the config saying plz delete all threat intel monitors before deleting the last config

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigResponse.java (hunk 4 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/action/SADeleteTIFSourceConfigResponse.java:22]: if the delete response is `OK` then the document is deleted successfully.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/common/Constants.java (hunk 5 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/common/Constants.java]: changed to `threat_intel_source_config_id`

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java (hunk 7 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java]: changed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java (hunk 6 above)
Comment (jowg-amazon) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestDeleteTIFSourceConfigAction.java]: fixed

**Comment by jowg-amazon** [comment]
> if there is only one config and it it's being deleted we need to ensure there are no jobs Fail the delete request of the config saying plz delete all threat intel monitors before deleting the last config

Added a method `checkAndEnsureThreatIntelMonitorsDeleted` that checks if the source config to be deleted is the last one and if so, makes sure that all threat intel monitors have been deleted. Otherwise it throws an exception.

**Review by eirsep — APPROVED** [review]

**Pull request merged**